### ISSUE TITLE
LLVM 17.x fails to build AdaptiveCpp (hipSYCL) for AMDGPU due to `invalid addrspacecast`

### ISSUE THREAD

**Issue by fodinabor** [llvm/llvm-project#68120]
For any of the current LLVM 17.x releases, [AdaptiveCpp](https://github.com/AdaptiveCpp/AdaptiveCpp) (hipSYCL) fails to build its test suit for AMDGPU, independent of the used ROCm version with the error `invalid addrspacecast`.
This can be found e.g. in the CI here: https://github.com/fodinabor/hipSYCL-LLVM-upstream-ci/actions/runs/6381943553/job/17319482716#step:12:569

I didn't take a closer look a the IR, yet.
[LLVM 16](https://github.com/AdaptiveCpp/AdaptiveCpp/actions/runs/6343017764/job/17229994676#step:14:14) and [upstream LLVM](https://github.com/fodinabor/hipSYCL-LLVM-upstream-ci/actions/runs/6378018187/job/17307792193) are fine.

I bisected both LLVM 16-17 and LLVM 17-main:
the first bad commit is: https://github.com/llvm/llvm-project/commit/72fc08a5412ec7ee7f0b904926db16cd86c1f876
the first good-again commit is: https://github.com/llvm/llvm-project/commit/d77c62053c944652846c00a35c921e14b43b1877

Applying d77c62053c944652846c00a35c921e14b43b1877 to `release/17.x` solves the issue, but not sure that's something we want to backport since it changes ABI..?

See the following scripts for reproducing it (I ran them in the following docker container: https://hub.docker.com/r/fodinabor/rocm-dev):
```bash
#!/bin/bash

# Use local dir as workdir
export WORKDIR=`pwd`

# Set ccache size to 512GB and use /tmp/ccache as the cache directory
export CCACHE_MAXSIZE=512G
export CCACHE_DIR=/tmp/ccache

# Clone the AdaptiveCpp repo
git clone https://github.com/AdaptiveCpp/AdaptiveCpp.git acpp

# Clone the llvm-project repository
git clone https://github.com/llvm/llvm-project.git

# Change into the llvm-project directory
cd llvm-project

# Start the bisect process
git bisect start

# Specify the good and bad commits
# git bisect old llvmorg-18-init
# git bisect new main
git bisect old llvmorg-17-init
git bisect new llvmorg-18-init

# Write a script to test the build
# This script should exit with code 0 if the build is successful
# and a non-zero code if the build fails
# For example:
#   ./build_and_test.sh

# Run the bisect process
git bisect run ~/build_and_test.sh

# Once the bisect process is complete, reset the repository to the original state
git bisect reset
```

build_and_test.sh
```bash
#!/bin/bash

rm -r $WORKDIR/install/*

mkdir -p build
cd build
cmake -G Ninja ../llvm -DLLVM_ENABLE_PROJECTS="clang;compiler-rt;lld;openmp" \
               -DOPENMP_ENABLE_LIBOMPTARGET=OFF \
               -DCMAKE_BUILD_TYPE=Release \
               -DCMAKE_INSTALL_PREFIX=$WORKDIR/install/llvm \
               -DLLVM_ENABLE_ASSERTIONS=OFF \
               -DLLVM_TARGETS_TO_BUILD="X86;AMDGPU" \
               -DCLANG_ANALYZER_ENABLE_Z3_SOLVER=0 \
               -DLLVM_INCLUDE_BENCHMARKS=0 \
               -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=ON \
               -DCMAKE_INSTALL_RPATH=$WORKDIR/install/llvm/lib \
               -DLLVM_ENABLE_OCAMLDOC=OFF \
               -DLLVM_ENABLE_BINDINGS=OFF \
               -DLLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=OFF \
               -DLLVM_ENABLE_DUMP=OFF \
               -DLLVM_CCACHE_BUILD=ON
if [ $? -ne 0 ]; then
    echo "LLVM CMake failed"
    exit 125
fi

ninja -j $OMP_NUM_THREADS install
if [ $? -ne 0 ]; then
    echo "LLVM install failed"
    exit 125
fi

cd $WORKDIR/acpp
rm -r build tests/build
mkdir -p build
cd build

cmake -G Ninja .. \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_INSTALL_PREFIX=$WORKDIR/install/acpp \
    -DLLVM_DIR=$WORKDIR/install/llvm/lib/cmake/llvm \
    -DWITH_ROCM_BACKEND=ON \
    -DWITH_CUDA_BACKEND=OFF \
    -DWITH_OPENCL_BACKEND=OFF \
    -DWITH_SSCP_COMPILER=OFF \
    -DWITH_STDPAR_COMPILER=OFF 
if [ $? -ne 0 ]; then
    echo "AdaptiveCpp CMake failed"
    exit 125
fi

ninja -j $OMP_NUM_THREADS install
if [ $? -ne 0 ]; then
    echo "Buliding AdaptiveCpp failed"
    exit 125
fi

cd $WORKDIR/acpp/tests
mkdir -p build
cd build
cmake -G Ninja .. \
    -DCMAKE_BUILD_TYPE=Release \
    -DAdaptiveCpp_DIR=$WORKDIR/install/acpp/lib/cmake/AdaptiveCpp \
    -DACPP_TARGETS="hip:gfx906"
if [ $? -ne 0 ]; then
    echo "AdaptiveCpp Tests CMake failed"
    exit 125
fi

ninja -j $OMP_NUM_THREADS |& tee build.log
if grep -q "error: invalid addrspacecast" build.log; then
    echo "AdaptiveCpp Tests build failed due to invalid addrspacecast"
    exit 1
fi
if [ $? -ne 0 ]; then
    echo "AdaptiveCpp Tests build failed"
    exit 125
fi

exit 0
```

cc @illuhad

**Comment by arsenm** [comment]
Can you attach the IR? The original IR is likely malformed in some way that happens to sometimes fold away

**Comment by fodinabor** [comment]
The IR: 
[reduction-hip-amdgcn-amd-amdhsa-gfx90c.ll.txt](https://github.com/llvm/llvm-project/files/12929929/reduction-hip-amdgcn-amd-amdhsa-gfx90c.ll.txt)

Note, I tried reproducing it from this IR - it only happens when loading our plugin.
More specifically, the issue only appears when adding `InstCombine` in `registerPipelineStartEPCallback`.
You can use the following example plugin that just runs InstCombine on that EP: [test_plugin.tar.gz](https://github.com/llvm/llvm-project/files/12930081/test_plugin.tar.gz)

I used the following to process this further: 
```
clang++ --rocm-device-lib-path=/opt/rocm/amdgcn/bitcode --rocm-path=/opt/rocm -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false --cuda-gpu-arch=gfx906 -fpass-plugin=test_plugin.so  -c reduction-hip-amdgcn-amd-amdhsa-gfx906.bc -target amdgcn-amd-amdhsa -O3
```

So, except if somebody with more ideas finds the underlying issue in the IR, we're hoping to be able to identify if the `registerPipelineStartEPCallback` is run for a PM that is targeting an (AMD) GPU. Since this extension is only actually required for CPUs in AdaptiveCpp, we would then just skip adding the passes. Any hints on how to identify this?

**Comment by arsenm** [comment]
I haven't managed to hit this error (I wish the invalid combination appeared in the error message, I'm guessing it's 4<->5 or 1<->5). I notice several puzzling things in your reproduction command. You're compiling the AMDGPU bitcode and using the superfluous --cuda-gpu-arch=gfx906, which is also a different subtarget than the bitcode is for. The attached is for gfx906, not gfx90c (not that it should matter for this).

I'm having trouble building the provided plugin:

```
 Undefined symbols:
  llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&), referenced from:
      llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) in test_plugin.cpp.o
  llvm::raw_ostream::write(unsigned char), referenced from:
      llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::printPipeline(llvm::raw_ostream&, llvm::function_ref<llvm::StringRef (llvm::StringRef)>) in test_plugin.cpp.o
  llvm::InstCombinePass::printPipeline(llvm::raw_ostream&, llvm::function_ref<llvm::StringRef (llvm::StringRef)>), referenced from:
      llvm::detail::PassModel<llvm::Function, llvm::InstCombinePass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::printPipeline(llvm::raw_ostream&, llvm::function_ref<llvm::StringRef (llvm::StringRef)>) in test_plugin.cpp.o

```

Can you try getting the net pass list with  -mllvm -print-pipeline-passes, and posting the IR from -O1 -Xclang -disable-llvm-passes?

**Comment by arsenm** [comment]
Also potentially interested, if I just run  opt -S -p=instcombine on the attached IR, I hit "LLVM ERROR: Instruction Combining did not reach a fixpoint after 1 iterations"

**Comment by fodinabor** [comment]
Ups, sorry for the confusion. I guess I ran the experiments on the server with 906, but then on my laptop trying to get the final reproducer used 90c.

`-mllvm -print-pipeline-passes` doesn't have an effect with `clang++` (cf https://discourse.llvm.org/t/clang-options-to-print-the-list-of-passes-thats-going-to-be-executed-for-a-compile-command-is-it-print-after-all/61032)
And with `opt` I haven't been able to reproduce this either.
See `-opt-bisect-limit` output: https://gist.github.com/fodinabor/3c8529b880fab44ba3a0c17bfe0f64f9

Regarding the plugin build issue... are you on MacOS or so? In AdaptiveCpp, we don't use the `add_llvm_pass_plugin` but link manually; possibly, adding the following helps?
```CMake
target_link_libraries(test_plugin
  ${LLVM_LIBS})
if(APPLE)
  target_link_libraries(test_plugin -Wl,-undefined -Wl,dynamic_lookup)
else()
  target_link_libraries(test_plugin -Wl,-znodelete)
endif()
```

I've been testing with LLVM commit 888437e1b60011b8a375dd30928ec925b448da57

An llvm-reduce version of the IR posted earlier (this was generated with `-save-temps`, but the reduced is the same for the `-disable-llvm-passes`).
Note, the reduced IR for the instcombine fixpoint error does not trigger the other issue down the line..
```llvm
target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-p7:160:256:256:32-p8:128:128-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7:8"
target triple = "amdgcn-amd-amdhsa"

%"class.hipsycl::glue::hiplike_dispatch::hiplike_reduction_descriptor.11" = type { i8, ptr, ptr, i32, [4 x i8], %"struct.hipsycl::sycl::detail::pointer_reduction_descriptor.12", ptr }
%"struct.hipsycl::sycl::detail::pointer_reduction_descriptor.12" = type <{ ptr, i32, %"struct.hipsycl::sycl::multiplies", [3 x i8] }>
%"struct.hipsycl::sycl::multiplies" = type { i8 }

; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: readwrite)
declare void @llvm.memcpy.p0.p4.i64(ptr noalias nocapture writeonly, ptr addrspace(4) noalias nocapture readonly, i64, i1 immarg) #0

; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: readwrite)
declare void @llvm.memcpy.p0.p0.i64(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i64, i1 immarg) #0

define amdgpu_kernel void @_Z16__hipsycl_kernelIZZZN7hipsycl4glue23hiplike_kernel_launcherILNS0_2rt10backend_idE1ENS3_9hip_queueEE4bindINS1_18kernel_name_traitsIN15reduction_tests16reduction_kernelIjZNS9_19test_two_reductionsIjEEvmmE7MultiOpLi213EEEZNSB_IjEEvmmEUlNS0_4sycl2idILi1EEERT_RT0_E_EELNS3_11kernel_typeE1ELi1ESL_JNSE_6detail28pointer_reduction_descriptorIjNSE_4plusIjEEEENSP_IjNSE_10multipliesIjEEEEEEEvNSF_IXT1_EEENSE_5rangeIXT1_EEESY_mT2_DpT3_ENUlPNS3_8dag_nodeEE_clES13_ENKUlDpT_E_clIJNS1_16hiplike_dispatch28hiplike_reduction_descriptorISS_NS19_15reduction_stageILi1EEEEENS1A_ISV_S1C_EEEEEDaS16_EUliDpRS15_E_Evv(ptr addrspace(4) byref(%"class.hipsycl::glue::hiplike_dispatch::hiplike_reduction_descriptor.11") align 8 %0) {
  %2 = alloca %"class.hipsycl::glue::hiplike_dispatch::hiplike_reduction_descriptor.11", align 8, addrspace(5)
  %3 = alloca %"class.hipsycl::glue::hiplike_dispatch::hiplike_reduction_descriptor.11", align 8, addrspace(5)
  %4 = addrspacecast ptr addrspace(5) %2 to ptr
  %5 = addrspacecast ptr addrspace(5) %3 to ptr
  call void @llvm.memcpy.p0.p4.i64(ptr %4, ptr addrspace(4) %0, i64 0, i1 false)
  call void @llvm.memcpy.p0.p0.i64(ptr %5, ptr %4, i64 0, i1 false)
  %6 = addrspacecast ptr %5 to ptr addrspace(5)
  call void @_ZN7hipsycl4glue16hiplike_dispatch37device_invocation_with_local_reducersIZNS1_42primitive_parallel_for_with_local_reducersI24__hipsycl_unnamed_kernelZZZNS0_23hiplike_kernel_launcherILNS_2rt10backend_idE1ENS6_9hip_queueEE4bindINS0_18kernel_name_traitsIN15reduction_tests16reduction_kernelIjZNSC_19test_two_reductionsIjEEvmmE7MultiOpLi213EEEZNSE_IjEEvmmEUlNS_4sycl2idILi1EEERT_RT0_E_EELNS6_11kernel_typeE1ELi1ESO_JNSH_6detail28pointer_reduction_descriptorIjNSH_4plusIjEEEENSS_IjNSH_10multipliesIjEEEEEEEvNSI_IXT1_EEENSH_5rangeIXT1_EEES11_mT2_DpT3_ENUlPNS6_8dag_nodeEE_clES16_ENKUlDpT_E_clIJNS1_28hiplike_reduction_descriptorISV_NS1_15reduction_stageILi1EEEEENS1C_ISY_S1E_EEEEEDaS19_EUliDpRS18_E_JS1F_S1G_EEEvSM_DpT1_EUlS1I_E_JS1F_S1G_EEEvOSK_DpT0_(ptr addrspace(5) %6)
  ret void
}

declare void @_ZN7hipsycl4glue16hiplike_dispatch37device_invocation_with_local_reducersIZNS1_42primitive_parallel_for_with_local_reducersI24__hipsycl_unnamed_kernelZZZNS0_23hiplike_kernel_launcherILNS_2rt10backend_idE1ENS6_9hip_queueEE4bindINS0_18kernel_name_traitsIN15reduction_tests16reduction_kernelIjZNSC_19test_two_reductionsIjEEvmmE7MultiOpLi213EEEZNSE_IjEEvmmEUlNS_4sycl2idILi1EEERT_RT0_E_EELNS6_11kernel_typeE1ELi1ESO_JNSH_6detail28pointer_reduction_descriptorIjNSH_4plusIjEEEENSS_IjNSH_10multipliesIjEEEEEEEvNSI_IXT1_EEENSH_5rangeIXT1_EEES11_mT2_DpT3_ENUlPNS6_8dag_nodeEE_clES16_ENKUlDpT_E_clIJNS1_28hiplike_reduction_descriptorISV_NS1_15reduction_stageILi1EEEEENS1C_ISY_S1E_EEEEEDaS19_EUliDpRS18_E_JS1F_S1G_EEEvSM_DpT1_EUlS1I_E_JS1F_S1G_EEEvOSK_DpT0_(ptr addrspace(5) byval(%"class.hipsycl::glue::hiplike_dispatch::hiplike_reduction_descriptor.11"))

attributes #0 = { nocallback nofree nounwind willreturn memory(argmem: readwrite) }
```

**Comment by ParkHanbum** [comment]
This problem seems to come from incomplete processing of the `addrspacecast` instruction.

In the first instcombine process, after the load instruction is removed as follows:
```
IC: Replacing   %i2 = load ptr, ptr %i1, align 8
    with ptr %arg
IC: Mod =   %i2 = load ptr, ptr %i1, align 8
    New =   %i2 = load ptr, ptr %i1, align 8
IC: ERASE   %i2 = load ptr, ptr %i1, align 8
ADD DEFERRED:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD DEFERRED:   store ptr %arg, ptr %i1, align 8
ADD:   store ptr %arg, ptr %i1, align 8
ADD:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
IC: Visiting:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
IC: Visiting:   store ptr %arg, ptr %i1, align 8
IC: Visiting:   store ptr %arg, ptr null, align 8
IC: Mod =   store ptr %arg, ptr null, align 8
    New =   store ptr poison, ptr null, align 8
ADD:   store ptr poison, ptr null, align 8
IC: Visiting:   store ptr poison, ptr null, align 8
IC: Visiting:   ret void
```
In the processing the first instruction `alloc` in the second instcombine process, `isAllocSiteRemovable` inside the `visitAllocSite` function is called.
```
ADD:   ret void
ADD:   store ptr poison, ptr null, align 8
ADD:   store ptr %arg, ptr %i1, align 8
ADD:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD:   %i = alloca ptr, align 8, addrspace(5)
IC: Visiting:   %i = alloca ptr, align 8, addrspace(5)
IC: Replacing   %i1 = addrspacecast ptr addrspace(5) %i to ptr
    with ptr poison
IC: ERASE   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD DEFERRED:   %i = alloca ptr, align 8, addrspace(5)
```
In this function, the `addrspacecast` and `store` instructions are removed.

This is the processing that should have occurred when the `load` instruction was removed. However, addrspacecast does not do this.

I will analyze it further and come up with a solution.

**Comment by ParkHanbum** [comment]
When addrspacecast is added to this DEFERRED list, can I also add it if the operand is alloc?

**Comment by ParkHanbum** [comment]
ok then I'll find other way.

**Comment by ParkHanbum** [comment]
based @arsenm  review, I summarized what is needed to do to solve this issue.

the `store`'s ptr is not only be `alloca`, but can be an Instruction such as `GEP` or `AddrSpaceCast`, which can be repeated multiple times, so `hasOneUse()` not to fit for this.

Search the users of the `store`'s ptr, (If there are commands such as `GEP` or `AddrSpaceCast` among the Users of the `store`'s ptr, these users must also be searched.)

In the `execution flow`,  check whether the `Store` has any affect when there is a `User` for Ptr after the store itself is executed. (ex, `Load` Inst). 

if removing `store` has no affect, we can remove it.

**Comment by arsenm** [comment]
> The IR: [reduction-hip-amdgcn-amd-amdhsa-gfx90c.ll.txt](https://github.com/llvm/llvm-project/files/12929929/reduction-hip-amdgcn-amd-amdhsa-gfx90c.ll.txt)
> 
> Note, I tried reproducing it from this IR - it only happens when loading our plugin. More specifically, the issue only appears when adding `InstCombine` in `registerPipelineStartEPCallback`. You can use the following example plugin that just runs InstCombine on that EP: [test_plugin.tar.gz](https://github.com/llvm/llvm-project/files/12930081/test_plugin.tar.gz)
> 
> I used the following to process this further:
> 
> ```
> clang++ --rocm-device-lib-path=/opt/rocm/amdgcn/bitcode --rocm-path=/opt/rocm -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false --cuda-gpu-arch=gfx906 -fpass-plugin=test_plugin.so  -c reduction-hip-amdgcn-amd-amdhsa-gfx906.bc -target amdgcn-amd-amdhsa -O3
> ```
> 

I was able to adjust this slightly and reproduce this with 17, but it works in 18 and main

**Comment by arsenm** [comment]
Modern opt is still producing the bad cast, and we're still emitting the error.

**Comment by arsenm** [comment]
Reduced (triple is necessary)
```
; RUN: opt -S -passes=instcombine %s | FileCheck %s

target triple = "amdgcn-amd-amdhsa"

define void @issue68120_invalid_addrspacecast_introduced(ptr addrspace(4) byref([56 x i8]) align 8 %arg) {
bb:
  %alloca = alloca [56 x i8], align 8, addrspace(5)
  %alloca1 = alloca [56 x i8], align 8, addrspace(5)
  call void @llvm.memcpy.p5.p4.i64(ptr addrspace(5) %alloca, ptr addrspace(4) %arg, i64 56, i1 false)
  call void @llvm.memcpy.p5.p5.i64(ptr addrspace(5) %alloca1, ptr addrspace(5) %alloca, i64 56, i1 false)
  %addrspacecast.alloca1 = addrspacecast ptr addrspace(5) %alloca1 to ptr
  %addrspacecast.0.to.5 = addrspacecast ptr %addrspacecast.alloca1 to ptr addrspace(5)
  call void @byval_func(ptr addrspace(5) %addrspacecast.0.to.5)
  ret void
}

declare void @byval_func(ptr addrspace(5) byval([56 x i8]))

declare void @llvm.memcpy.p0.p4.i64(ptr noalias nocapture writeonly, ptr addrspace(4) noalias nocapture readonly, i64, i1 immarg) #0
declare void @llvm.memcpy.p0.p0.i64(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i64, i1 immarg) #0
declare void @llvm.memcpy.p5.p4.i64(ptr addrspace(5) noalias nocapture writeonly, ptr addrspace(4) noalias nocapture readonly, i64, i1 immarg) #0
declare void @llvm.memcpy.p5.p5.i64(ptr addrspace(5) noalias nocapture writeonly, ptr addrspace(5) noalias nocapture readonly, i64, i1 immarg) #0

attributes #0 = { nocallback nofree nounwind willreturn memory(argmem: readwrite) }

```

**Comment by arsenm** [comment]
InstCombine is checking isValidAddrSpaceCast, but incorrectly in this context. For this case, it should be checking if this is a no-op addrspacecast, not a valid one. But in this situation, it isn't even calling this for the cast that ultimately is introduced

**Comment by arsenm** [comment]
Unlike the other cases,  AddrSpaceCast wasn't adding users to the PointerReplacer worklist. Fixing this avoids this issue, but breaks some other tests I need to investigate 

```
@@ -342,6 +342,8 @@ bool PointerReplacer::collectUsersRecursive(Instruction &I) {
       Worklist.insert(Inst);
     } else if (isEqualOrValidAddrSpaceCast(Inst, FromAS)) {
       Worklist.insert(Inst);
+      if (!collectUsersRecursive(*Inst))
+        return false;
     } else if (Inst->isLifetimeStartOrEnd()) {
       continue;
     } else {

```
